### DISCUSSION TITLE
How to correctly delete family member provider?

### DISCUSSION THREAD

**Discussion by vorant94** [rrousselGit/riverpod#2142] [Q&A] [answered]
Hi, there

I'm new to riverpod and need some help. Let's say I have Todos state like so

```dart
@riverpod
class Todos extends _$Todos {
  @override
  List<Todo> build() {
    return [];
  }

  // ...

  Future<void> delete(int id) async {
    // ...
  }
}
```

And also I have family provider to get todo by its id:
```dart
@riverpod
Todo todo(TodoRef ref, Id todoId) {
  return ref
      .watch(todosProvider)
      .firstWhere((todo) => todo.id == todoId);
}
``` 

At last here is TodoScreen where I listen for a particular Todo to show it and allow to delete it
```dart
  @override
  Widget build(BuildContext context) {
    final todo = ref.watch(todoProvider(widget.todoId));

    return Scaffold(
      // ...
    );
  }

  void _deleteTodo() {
    final notifier = ref.read(todosProvider.notifier);
    notifier.delete(widget.todoId);
    context.pop()
  }
```

The problem is when I delete Todo I get `Unhandled Exception: Bad state: No element` exception in `todoProvider` (it fails in `firstWhere`). Trying to pop of the context before deletion doesn't help. How should I properly stop watching and delete a family member provider? I don't want to make `todoProvider` to return `Todo?` because it always return todo, there is no todo to return in this single case with deletion a moment before widget that is listening for a provider is destroyed

**Comment by AhmedLSayed9** [comment, marked as answer]
You can return current state (Todo) when your `todosProvider` delete it:
```
@riverpod
Todo todo(TodoRef ref, Id todoId) {
  return ref
      .watch(todosProvider)
      .firstWhereOrNull((todo) => todo.id == todoId) ?? ref.state;
}
```
Be careful that if your `todosProvider` initially doesn't have your todo, ref.state will throw as there's no state yet, but I guess that's not happening in your case.

**Reply by vorant94** [reply]
it results in `Bad state: Tried to read the state of an uninitialized provider`🥲

the only way the user can get to TodoScreen is by clicking on one of todos from the list on previous screen, so with the exception for delete case there is no way `todosProvider` doesn't have target todo

**Reply by AhmedLSayed9** [reply]
Ahh `ref.state` loses prev state when the provider rebuild due to ref.watch and it needs the setter to be called first, in that case it's useless here.

You can use ref.listen and imperatively update state only if it the todo exists in the todos list:
```
final todoProvider = Provider.family<Todo, String>((ref, todoId) {
  ref.listen<List<Todo>>(
    todosProvider,
    (previous, next) {
      if (next.any((t) => t.id == todoId)) {
        ref.state = next.firstWhere((t) => t.id == todoId);
      }
    },
    fireImmediately: true,
  );
  return ref.state;
});
```

**Reply by vorant94** [reply]
thanks!
